pkg/didcomm/protocol/didexchange/service.go

```diff
@@ -63,21 +112,115 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+		eventsTriggered, err = s.sendEvent(msg, thid, next)
+		if err != nil {
+			return fmt.Errorf("send events failed: %w", err)
+		}
+	}
+
+	// if no events are trigger continue the execution
+	if !eventsTriggered {
+		return s.handle(&message{Msg: msg, ThreadID: thid, NextStateName: next.Name()})
+	}
+
+	return nil
+}
+
+// RegisterActionEvent on DID Exchange protocol messages. The events are triggered for incoming ConnectionRequest,
+// ConnectionResponse or ConnectionAck message types. The consumer need to invoke the callback to resume processing.
+// Only one channel can be registered for the action events. If called multiple times, the events will be sent to the
+// last channel. This works in conjunction with RegisterAutoExecute(). If this function is called after
+// RegisterAutoExecute(), the service will not execute the processing automatically and it'll wait for the callback. If
+// both are not set, then service will not handle the messages and throw an error.
+func (s *Service) RegisterActionEvent(ch chan<- event.DIDCommEvent) error {
+	s.lock.Lock()
+	s.actionEvent = ch
+	s.execute = true
+	s.lock.Unlock()
+
+	return nil
+}
+
+// UnregisterActionEvent on DID Exchange protocol messages. Refer RegisterActionEvent().
+func (s *Service) UnregisterActionEvent() error {
+	return s.disableExecute()
+}
+
+// RegisterStreamEvent on DID Exchange protocol messages. The events are triggered for incoming ConnectionRequest,
+// ConnectionResponse or ConnectionAck message types. The Callback is set to nil in the actionEvent message and service
+// won't be expecting any response from the consumers.
+func (s *Service) RegisterStreamEvent(ch chan<- dispatcher.DIDCommMsg) error {
+	s.streamLock.Lock()
+	s.streamEvents = append(s.streamEvents, ch)
+	s.streamLock.Unlock()
+
+	return nil
+}
+
+// UnregisterStreamEvent on DID Exchange protocol messages. Refer RegisterStreamEvent().
+func (s *Service) UnregisterStreamEvent(ch chan<- dispatcher.DIDCommMsg) error {
+	s.streamLock.Lock()
+	for i := 0; i < len(s.streamEvents); i++ {
+		if s.streamEvents[i] == ch {
+			s.streamEvents = append(s.streamEvents[:i], s.streamEvents[i+1:]...)
+			i--
+		}
+	}
+	s.streamLock.Unlock()
+
+	return nil
+}
+
+// RegisterAutoExecute on DID Exchange protocol messages. When this function is called, the service will auto execute
+// the workflow. This works in conjunction with RegisterActionEvent(). If this function is called after
+// RegisterActionEvent(), the service will execute the processing automatically and no action events will be triggered.
+// If both are not set, then service will not handle the messages and throw an error.
+func (s *Service) RegisterAutoExecute() error {
+	s.lock.Lock()
+	s.execute = true
+	s.lock.Unlock()
+
+	return nil
+}
+
+// UnregisterAutoExecute on DID Exchange protocol messages. Refer RegisterAutoExecute().
+func (s *Service) UnregisterAutoExecute() error {
```

Comment:
This should not disable the events stream.

Reply:
I assume you mean DIDCommMsg as a result of RegisterStreamEvent :). Agree those should not be disabled.

Reply:
@llorllale @troyronda Is the code missing something here? Does this need any change ?